### PR TITLE
drivers: sensor: a01nyub: added driver

### PR DESCRIPTION
**Pull request by AnachronisticPenguin**
Added a driver for the DFRobot A01NYUB distance sensor. This sensor sends its readings via UART at 9600 baud. This driver uses interrupts to read the data from the sensor.

### CODE REVIEW

File: drivers/sensor/Kconfig [hunk 1]
```diff
@@ -57,6 +57,8 @@ config SENSOR_INFO
 
 comment "Device Drivers"
 
+source "drivers/sensor/a01nyub/Kconfig"
+
```
Comment (MaureenHelm) [inline drivers/sensor/Kconfig:61]: Please remove the empty line

File: drivers/sensor/a01nyub/a01nyub.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	/* Max data length is 16 bits */
+	uint16_t data;
+	uint8_t xfer_bytes;
+	uint8_t rd_data[A01NYUB_BUF_LEN];
+};
+
+struct a01nyub_cfg {
+	const struct device *uart_dev;
+	uart_irq_callback_user_data_t cb;
+};
```
Comment (MaureenHelm) [inline drivers/sensor/a01nyub/a01nyub.h:36]: This isn't enough to justify a separate private header file. Please move these lines to `a01nyub.c`

File: samples/sensor/a01nyub/src/main.c [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		LOG_ERR("sensor: device not found.");
+		return 0;
+	}
+
+	while (1) {
+		if (sensor_sample_fetch(dev) != 0) {
+			LOG_ERR("sensor: sample fetch fail.");
+		}
+
+		if (sensor_channel_get(dev, SENSOR_CHAN_DISTANCE, &val) != 0) {
```
Comment (MaureenHelm) [inline samples/sensor/a01nyub/src/main.c:32]: We're trying to reduce the number of driver-specific samples. Please abstract this sample to allow support for other `SENSOR_CHAN_DISTANCE` sensor drivers. See `samples/sensor/*_polling` for example

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thank you for the contribution!

File: samples/sensor/a01nyub/src/main.c (hunk 3 above)
Comment (AnachronisticPenguin) [reply samples/sensor/a01nyub/src/main.c]: That's good guidance, thanks. In this case, the sample doesn't add anything that is not already covered by the mhz19b sample, which is another interrupt-driven UART-based sensor. I have removed it from the PR.

File: drivers/sensor/a01nyub/a01nyub.h (hunk 2 above)
Comment (AnachronisticPenguin) [reply drivers/sensor/a01nyub/a01nyub.h]: Done!

File: drivers/sensor/Kconfig (hunk 1 above)
Comment (AnachronisticPenguin) [reply drivers/sensor/Kconfig]: Done!

File: drivers/sensor/CMakeLists.txt [hunk 4]
```diff
@@ -1,12 +1,13 @@
 # SPDX-License-Identifier: Apache-2.0
 
+add_subdirectory_ifdef(CONFIG_A01NYUB		a01nyub)
```
[the remark this answers is not in the document]
Comment (AnachronisticPenguin) [reply drivers/sensor/CMakeLists.txt]: Done!

File: drivers/sensor/CMakeLists.txt [hunk 5]
```diff
@@ -1,12 +1,13 @@
...
-add_subdirectory_ifdef(CONFIG_ADXL345 adxl345)
-add_subdirectory_ifdef(CONFIG_ADXL362 adxl362)
-add_subdirectory_ifdef(CONFIG_ADXL372 adxl372)
-add_subdirectory_ifdef(CONFIG_AK8975 ak8975)
+add_subdirectory_ifdef(CONFIG_ADT7310		adt7310)
+add_subdirectory_ifdef(CONFIG_ADT7420		adt7420)
+add_subdirectory_ifdef(CONFIG_ADXL345		adxl345)
+add_subdirectory_ifdef(CONFIG_ADXL362		adxl362)
+add_subdirectory_ifdef(CONFIG_ADXL372		adxl372)
+add_subdirectory_ifdef(CONFIG_AK8975		ak8975)
```
[the remark this answers is not in the document]
Comment (AnachronisticPenguin) [reply drivers/sensor/CMakeLists.txt]: Not entirely sure how git decided that I wanted to achieve adding a new line by making these unrelated changes. There was a rebase to main in the midst of this whole process, which probably has something to do with it. Regardless, it is fixed.

**Review by MaureenHelm — COMMENTED** [review]
One more thing - can you add this sensor to `tests/drivers/build_all/sensor/uart.dtsi`? This will ensure it builds in CI

**Review by MaureenHelm — APPROVED** [review]

**Comment by AnachronisticPenguin** [comment]
@nashif , @galak : bringing this PR back to the top of your inbox now that the merge window has reopened.  Please let me know if any further edits are required.

**Review by carlescufi — APPROVED** [review]

**Pull request merged**